src/utils/src/ExponentialBackoffRetryStrategy.c

```diff
@@ -0,0 +1,251 @@
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PExponentialBackoffRetryStrategyState pExponentialBackoffRetryStrategyState = NULL;
+    PExponentialBackoffRetryStrategyConfig pExponentialBackoffConfig = NULL;
+
+    CHK(ppRetryStrategy != NULL, STATUS_NULL_ARG);
+
+    // If no config provided, create retry strategy with default config
+    if (pRetryStrategyConfig == NULL) {
+        return exponentialBackoffRetryStrategyWithDefaultConfigCreate(ppRetryStrategy);
+    }
+
+    pExponentialBackoffConfig = TO_EXPONENTIAL_BACKOFF_CONFIG(pRetryStrategyConfig);
+    CHK_STATUS(validateExponentialBackoffConfig(pExponentialBackoffConfig));
+
+    pExponentialBackoffRetryStrategyState = (PExponentialBackoffRetryStrategyState) MEMCALLOC(1, SIZEOF(ExponentialBackoffRetryStrategyState));
+    CHK(pExponentialBackoffRetryStrategyState != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    pExponentialBackoffRetryStrategyState->exponentialBackoffRetryStrategyConfig = *pExponentialBackoffConfig;
+    CHK_STATUS(resetExponentialBackoffRetryState(pExponentialBackoffRetryStrategyState));
+    DLOGD("Created exponential backoff retry strategy state with provided retry configuration.");
+
+CleanUp:
+    if (STATUS_SUCCEEDED(retStatus)) {
+        (*ppRetryStrategy) = (PRetryStrategy)pExponentialBackoffRetryStrategyState;
+    } else {
+        SAFE_MEMFREE(pExponentialBackoffRetryStrategyState);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+UINT64 getRandomJitter(UINT32 jitterFactor) {
+    return (RAND() % jitterFactor) * HUNDREDS_OF_NANOS_IN_A_MILLISECOND;
+}
+
+STATUS calculateWaitTime(PExponentialBackoffRetryStrategyState pRetryState, PExponentialBackoffRetryStrategyConfig pRetryConfig, PUINT64 pWaitTime) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT64 power = 0, waitTime = 0;
+
+    CHK(pWaitTime != NULL, STATUS_NULL_ARG);
+
+    CHK_STATUS(computePower(DEFAULT_KVS_EXPONENTIAL_FACTOR, pRetryState->currentRetryCount, &power));
+    waitTime = power * pRetryConfig->retryFactorTime;
+
+    *pWaitTime = waitTime;
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS validateAndUpdateExponentialBackoffStatus(PExponentialBackoffRetryStrategyState pExponentialBackoffRetryStrategyState) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+
+    switch (pExponentialBackoffRetryStrategyState->status) {
+        case BACKOFF_NOT_STARTED:
+            pExponentialBackoffRetryStrategyState->status = BACKOFF_IN_PROGRESS;
+            DLOGD("Status changed from BACKOFF_NOT_STARTED to BACKOFF_IN_PROGRESS");
+            break;
+        case BACKOFF_IN_PROGRESS:
+            // Do nothing
+            DLOGD("Current status is BACKOFF_IN_PROGRESS");
+            break;
+        case BACKOFF_TERMINATED:
+            DLOGD("Cannot execute exponentialBackoffBlockingWait. Current status is BACKOFF_TERMINATED");
+            CHK(FALSE, STATUS_EXPONENTIAL_BACKOFF_INVALID_STATE);
```

Comment:
You could use CHK_ERR in these cases to output the text in the error level if you want

Reply:
done